### DISCUSSION TITLE
Additional eturnal whitelist option

### DISCUSSION THREAD

**Discussion by daniele-barresi** [nextcloud/all-in-one#3395] [Ideas] [answered]
Hi there,
I'm using NextcloudAIO and it really makes it easy to install and maintain, so big thanks for the huge efforts!

I'm currently using Nextcloud on premise, behind a NAT. I know it's not fully recommended for Talk STUN/TURN setup, but I believe it works with the right config on NAT, firewall and DNS. In my opinion it's an ok setup when most of the time the call participants are actually on premise, and occasionally some participants connect from the Internet. 

The problem I'm currently facing is with TURN server for call participants that are "inside" of the network, meaning they are connecting behind the same firewall as the Nextcloud server (although on different subnets). I have found it works properly when I add the LAN network CIDR in the "whitelist" section of eturnal configuration file, to override the corresponding `blacklist: -recommended` setting, which implicitly includes all local networks.
I did so by adding a line in the `whitelist` section of `/conf/eturnal.ym` file on the running nextcloud-aio-talk container and reloading with 'eturnalctl reload`. This then allows local clients to establish a connection with the TURN server. 
This configuration, off course, wont survive a restart of the docker container.

In my scenario, it would be very useful to have an additional line in the `/containers/talk/start.sh` script to include an environment variable to define an additional whitelist value. Another option could be to mount the start.sh script from a configuration file, so that it could be expanded if necessary.

I'm also open to know if I'm missing any alternative solution to add a whitelist option to eturnal in the nextcloud-aio-talk container.

Cheers

**Comment by szaimen** [comment]
Hi, I am wondering since when is this necessary? Since we switchted to eturnal? 

How long are you using AIO already? 

---

> I'm using NextcloudAIO and it really makes it easy to install and maintain, so big thanks for the huge efforts!

Thanks, this is great to hear!

**Reply by daniele-barresi** [reply]
I started using AIO recently, I migrated from snap package. 
I never used the TURN server before, as Talk was not being used at all in the organization.
We are currently testing Talk as a more private alternative to other popular video conferencing services. It also integrates nicely with the calendar app, which we have been using for a while now.

With the current configuration of the nextcloud-aio-talk container, `start.sh` includes the `recommended` option in blacklist for eturnal, which implicitly blocks every private network address space. This is why eturnal will refuse to respond to participants in the local network, as they are behind the same firewall/router as the Nextcloud server, so they connect tu eturnal with local IPs (no NAT involved). I'm not sure when TURN server was switched to eturnal, but I guess the same applied to coTurn as well.

I understand the default is to block local network addresses, as there are security concerns about it (https://www.rtcsec.com/article/cve-2020-26262-bypass-of-coturns-access-control-protection/#further-concerns-what-else). Still, it would be good to have the option to configure a network as whitelisted for specific use cases.

**Comment by daniele-barresi** [comment]
> Hi, I am wondering since when is this necessary? Since we switchted to eturnal?

I was checking the Dockerfile for Talk container in Nextcloud AIO repository. In 7.0.0 it uses `strukturag/nextcloud-spreed-signaling:1.1.3`, which I believe uses coTurn. Since 7.1.0 the Dockerfile also includes `eturnal/eturnal:1.11.1`. 

I believe the adoption of eTurnal and the default blacklist for all private network addresses is very likely the cause of the issues of Talk calls with Nextcloud behind the same firewall as the call participants, as others are reporting in #3350

I also added more information on my current setup (with Talk calls working) here:
https://github.com/nextcloud/all-in-one/discussions/3350#discussioncomment-7111109

**Reply by szaimen** [reply]
What I find so weird is that we had blacklisting of private ip ranges in coturn before eturnal: https://github.com/nextcloud/all-in-one/commit/aeae72f5ba24d186d0e9b83ecc1e3e4387163949#diff-2251ec8a5310d7028953402eed81f74318fd30a3f3e0b7d3775153fb077ca47e

So maybe with coturn it simply didnt work and now that we moved to eturnal it made the configuration work?

**Reply by sando38** [reply]
Hello all, hello @szaimen ,

just came across this discussion.

We checked some comments with @weiss and it maybe that some setups ran into the blacklist denial rules - as some admins have found out already.

In TURN we have to distinguish between clients and peers. Clients are basically those that request a TURN relay which then may be consumed by a peer. The peer is the addressee of a communication which should be [protected](https://www.rtcsec.com/article/cve-2020-26262-bypass-of-coturns-access-control-protection/#further-concerns-what-else). The client side authenticates itself usually by the authentication mechanisms, like with ephemeral credentials. So this could be a sufficient protection already.

Eturnal's behaviour is to address both clients and peers with blacklist and whitelists. This may not be an optimal behaviour for some use cases applied with Nextcloud AIO, but it worked until now in the environments where eturnal has been successfully deployed. 

The issues observed seem to be with clients connecting to eturnal from internal/ local network addresses which will be blocked by eturnal's blacklist. The workaround would be like described in this discussion.

We may provide a quick fix for the separation of client/ peer which was on [our list](https://github.com/processone/eturnal/issues/47) anyway.

Cheers,
saarko

**Reply by szaimen** [reply]
Hi @sando38, glad to see you here! 👋

> We may provide a quick fix for the separation of client/ peer which was on [our list](https://github.com/processone/eturnal/issues/47) anyway.

This would be amazing so that we dont need to add a workaround on our side :)

**Reply by daniele-barresi** [reply]
> In TURN we have to distinguish between clients and peers. Clients are basically those that request a TURN relay which then may be consumed by a peer. The peer is the addressee of a communication which should be [protected](https://www.rtcsec.com/article/cve-2020-26262-bypass-of-coturns-access-control-protection/#further-concerns-what-else). The client side authenticates itself usually by the authentication mechanisms, like with ephemeral credentials. So this could be a sufficient protection already.

Thank's for this clarification. It makes a lot more sense to me now.

> The issues observed seem to be with clients connecting to eturnal from internal/ local network addresses which will be blocked by eturnal's blacklist. The workaround would be like described in this discussion.

I can confirm that the workaround works indeed. In my case, I added a cronjob on Nextcloud server that will periodically check if the local network is whitelisted in eturnal's configuration. In case it's not, the script adds the whitelist and reloads eturnal. It's an attempt to make this configuration as "persistent" as possibile, by making the change in the running Docker container, but without tampering with the image itself, so the config will survive a reboot or an upgrade.

> We may provide a quick fix for the separation of client/ peer which was on [our list](https://github.com/processone/eturnal/issues/47) anyway.

This is great news! Thank you for the effort!
It will be the best solution, no workarounds needed.
🙏 👏

**Reply by szaimen** [reply]
PR is here: https://github.com/nextcloud/all-in-one/pull/3426

**Reply by tuetenk0pp** [reply]
I see the PR has been merged. I am in the middle of setting up a new instance and I think I ran into this issue as well. How long until the next release? Just so that I know how long before I can migrate our users to the new instance.

**Reply by szaimen** [reply]
I let you know as soon as a new beta release is available that includes the fix

**Comment by szaimen** [comment, marked as answer]
This is now fixed with v7.4.0 Beta. Testing and feedback is welcome! See https://github.com/nextcloud/all-in-one#how-to-switch-the-channel

**Reply by sol8712** [reply]
Very nice! I'm switching now and will report any issues. Thank you for all your work!

**Reply by ralfrupf1976** [reply]
I just updated to v7.4.1 - and YES - the connection test doesn't fail anymore!! Thank you

**Reply by daniele-barresi** [reply]
I just updated to NC AIO 7.5.1, left default black/whitelisting in /conf/eturnal.yml and everything works smoothly!

Thank you again!

**Reply by ollioddi** [reply]
Hi @szaimen.

I am running v7.12.1 and is still facing this issue.

```
/conf # cat eturnal.yml
eturnal:
  listen:
    - ip: "::"
      port: 3478
      transport: udp
    - ip: "::"
      port: 3478
      transport: tcp
  log_dir: stdout
  log_level: warning
  secret: "verysecret"
  relay_ipv4_addr: "172.20.0.2"
  blacklist_peers:
  - recommended
  whitelist_peers:
  - 127.0.0.1
  - ::1
  - "172.20.0.2"
```
  
  I simply cannot use it locally, on `192.168.0.0/24`. It seems **everything** is read-only or reset upon restart. Do you have any suggestions? Otherwise i might need a new discussion?
  
  Thanks in advance.

**Reply by apstrom** [reply]
I am having this exact issue. I cannot edit the eturnal.conf to adjust the peers such that local addresses may access the server.

**Comment by sol8712** [comment]
When i get home from work in about 5 hours i will test this on my local install and report back.

-------- Original Message --------

**Comment by seanmuth** [comment]
Hi again @szaimen  😉 

Now that I have my internal AIO turn server working, I'd also like to use it with Matrix, hosted on the same server (but different docker network)

If I add the CIDR for the matrix docker network subnet to eturnal's whitelist, and `eturnalclt reload` that allows the matrix clients to use the turn server as expected, but this config doesn't survive a restart. Is there a way to pass eturnal configs via AIO mastercontainer's docker-compose.yml file, similar to the talk port?

**Reply by szaimen** [reply]
I fear the included turn server is only intended to be used with Nextcloud talk and not with any external service. If you want that you need to set up your own turn server externally from AIO.

**Reply by seanmuth** [reply]
Understandable, it's definitely outside the scope of AIO. 

I should be able to use a solution like @daniele-barresi 's that just makes sure the additional docker network is whitelisted, add it if not and reload eturnal.

Will also just try running coturn or eturnal on my host server so i have full control of the configs
